rpc/rest/security/rest_security.go

```diff
@@ -0,0 +1,304 @@
+		a.groupDb = append(a.groupDb, newPermissionGroup)
+	}
+}
+
+// Validate the request
+func (a *authenticator) Validate(provider http.HandlerFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		authHeader := req.Header.Get(authHeaderStr)
+		if authHeader == "" {
+			a.formatter.Text(w, http.StatusUnauthorized, "401 Unauthorized: authorization header required")
+			return
+		}
+		bearerToken := strings.Split(authHeader, " ")
+		if len(bearerToken) != 2 {
+			a.formatter.Text(w, http.StatusUnauthorized, "401 Unauthorized: invalid authorization token")
+			return
+		}
+		token, err := jwt.Parse(bearerToken[1], func(token *jwt.Token) (interface{}, error) {
+			if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+				return nil, fmt.Errorf("error parsing token")
+			}
+			return []byte(secret), nil
+		})
+		if err != nil {
+			errStr := fmt.Sprintf("500 internal server error: %s", err)
+			a.formatter.Text(w, http.StatusInternalServerError, errStr)
+			return
+		}
+
+		if err := a.validateToken(token.Raw, req.URL.Path); err != nil {
+			errStr := fmt.Sprintf("401 Unauthorized: %v", err)
+			a.formatter.Text(w, http.StatusUnauthorized, errStr)
+			return
+		}
+
+		provider.ServeHTTP(w, req)
+	})
+}
+
+// Register authenticator-wide security handlers
+func (a *authenticator) registerSecurityHandlers() {
+	a.router.HandleFunc(login, a.createTokenEndpoint).Methods(http.MethodPost)
+	a.router.HandleFunc(logout, a.invalidateTokenEndpoint).Methods(http.MethodPost)
+}
+
+// Validates credentials and provides new token
+func (a *authenticator) createTokenEndpoint(w http.ResponseWriter, req *http.Request) {
+	name, errCode, err := a.validateCredentials(req)
+	if err != nil {
+		a.formatter.Text(w, errCode, err.Error())
+		return
+	}
+	var claims jwt.MapClaims = make(map[string]interface{})
+	if a.expTime != 0 {
+		claims["exp"] = time.Now().Add(time.Minute * time.Duration(a.expTime)).Unix()
+	}
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+	tokenString, err := token.SignedString([]byte(secret))
+	if err != nil {
+		errStr := fmt.Sprintf("500 internal server error: faield to sign token: %v", err)
+		a.formatter.Text(w, http.StatusInternalServerError, errStr)
+		return
+	}
+	a.tokenDb[name] = tokenString
+	a.formatter.Text(w, http.StatusOK, tokenString)
+	return
+}
+
+// Removes token endpoint from the DB. During processing, token will not be found and will be considered as invalid.
+func (a *authenticator) invalidateTokenEndpoint(w http.ResponseWriter, req *http.Request) {
+	name, _, ok := req.BasicAuth()
+	if !ok {
+		a.formatter.Text(w, http.StatusInternalServerError, "500 internal server error: unable to process authentication data")
+		return
+	}
+	delete(a.tokenDb, name)
+}
+
+// Validates credentials, returns name and error code/message if invalid
+func (a *authenticator) validateCredentials(req *http.Request) (string, int, error) {
```

Comment:
If you are gonna extract this function than it should just check user/password combination and not care about request at all, otherwise it doesn't really help to extract it, because it's coupled with http stuff.

Reply:
will be fixed in following update